LinearStretching enhancements

johnnychen94 opened the pull request:
It's just a prototype for #27 , does this usage look good to you? 

```julia
img = float.(testimage("mandril_gray"))
imgo = adjust_histogram(img, LinearStretching(src_minval=0.1, src_maxval=0.9))
```

Note that this implementation doesn't always work for `N0f8` types, there're two ways to work this around: clamp01 or promote storage type. I'm not sure which is better.

cc: @zygmuntszpak @bjarthur

johnnychen94 committed:
LinearStretching enhancements

* expose src_minval and src_maxval of LinearStretching ( issue #27)
* performance tweak by
  * avoiding recomputation in inner loop
  * early return for trivial case
* clamp values to (minval, maxval)

bjarthur commented:
great!  in what cases doesn't it work for N0f8?

johnnychen94 commented:
> in what cases doesn't it work for N0f8?

The kernel operation of `linearstrech` can be reduced to

```julia
out .= (b-a)/(B-A) * (X - (A*b-B*a)/(b-a))
```

where `(A, B)` are `(src_minval, src_maxval)` and `(a, b)` are `(minval, maxval)`.

This expression says if `finitemin(X) < (A*b-B*a)/(b-a)`, the output exceeds the `[0, 1]` range.

```
julia> imgo = adjust_histogram(img, LinearStretching(src_minval=0.1, src_maxval=0.9))
ERROR: ArgumentError: element type FixedPointNumbers.Normed{UInt8,8} is an 8-bit type representing 256 values from 0.0 to 1.0,
  but the values (-0.0024509803921568714,) do not lie within this range.
  See the READMEs for FixedPointNumbers and ColorTypes for more information.
```

Paremeters `(src_minval, src_maxval)` and `(minval, maxval)` are a little bit mutually exclusive for N0f8 types.

The simplest solution is to do `clamp01` on the results, but that can be a breaking behavior for float types; not to mention the concept `linear_stretch` doesn't imply `clamp01`. Thus I don't like this solution.

My opinion is to:

* if the user does not specify parameter, use default values: `(0., 1., minfinite(img), maxfinite(img))`.
* if the user specifies either `from` or `to`, infer the other to make sure output value doesn't exceed range `[0, 1]`. -- this might not so trivial though...
* if the user specifies both, throw a warning when the condition doesn't satisfied for N0f8 types. And let the user recheck the parameters or promote eltype to floats.

bjarthur commented:
this issue should exist for all fixed-point types not just N0f8 IIUC.

`imadjustintensity` clamped the output to [0,1].  this precedent is/was intuitive to me.

however, i see your point.  give the user the flexibility to stretch outside [0,1] and then clamp themselves if that is what they desire.  and if they stretch into a type which can't support values outside [0,1], then either throw an error or promote.

maybe make clamping easy for them by having a `clamp=false` keyword arg?

johnnychen94 commented:
Recent changes (including 0572eb4):

* performance tweak
* `clamp` wrt (minval, maxval) -- This solves FixedPoint Issue

Performance tweak get ~10x speed up:

```julia
julia> img = testimage("mandril_gray");
julia> alg = LinearStretching();

# master
julia> @benchmark adjust_histogram($img, $alg)
BenchmarkTools.Trial:
  memory estimate:  256.19 KiB
  allocs estimate:  3
  --------------
  minimum time:     4.473 ms (0.00% GC)
  median time:      4.739 ms (0.00% GC)
  mean time:        4.804 ms (0.03% GC)
  maximum time:     7.016 ms (0.00% GC)
  --------------
  samples:          1041
  evals/sample:     1

# this PR
julia> @benchmark adjust_histogram($img, $alg)
BenchmarkTools.Trial:
  memory estimate:  256.14 KiB
  allocs estimate:  2
  --------------
  minimum time:     493.738 μs (0.00% GC)
  median time:      535.095 μs (0.00% GC)
  mean time:        550.216 μs (0.36% GC)
  maximum time:     2.837 ms (80.10% GC)
  --------------
  samples:          9068
  evals/sample:     1
```

More benchmark on this PR:

```julia
julia> img = testimage("mandril_gray");

julia> alg = LinearStretching();

julia> @btime adjust_histogram($img, $alg);
  483.575 μs (2 allocations: 256.14 KiB)

julia> @btime adjust_histogram($(float.(img)), $alg);
  880.483 μs (2 allocations: 1.00 MiB)

julia> alg = LinearStretching(src_minval=0.1, src_maxval=0.9);

# clamp contributes ~200μs for N0f8
julia> @btime adjust_histogram($img, $alg);
  670.129 μs (2 allocations: 256.14 KiB)

# but doesn't for Float32
julia> @btime adjust_histogram($(float.(img)), $alg);
  894.220 μs (2 allocations: 1.00 MiB)

johnnychen94 commented on src/algorithms/linear_stretching.jl:
```diff
@@ -59,24 +59,45 @@ imgo = adjust_histogram(img, LinearStretching(minval = 0, maxval = 1))
-            return  T <: Integer ? round(Int, newval ) : newval
+            # equivalent to `(x-A) * ((b-a)/(B-A)) + a` where
+            # (A, B, a, b) are (src_minval, src_maxval, minval, maxval)
+            newval = r * img[p] - o
```
```julia
newval = clamp(r * img[p] -o, minval, maxval)
```
would fix the fixedpoint issue, but the running time is 2~3x than this version.

Edit:

looks like the benchmark was polluted by Atom and/or Revise.

johnnychen94 committed:
ImageContrastAdjustment v0.3.5

johnnychen94 commented:
@zygmuntszpak I think this PR is ready for review/merge, commits are rebased into two.

Summary:

* expose src_minval and src_maxval of LinearStretching ( issue #27)
* performance tweak by
  * avoiding recomputation in inner loop ( before: __5ms__ after: __500 μs__)
  * early return for trivial case ( before: __500 μs__ after: __22 μs__)
* clamp values to (minval, maxval)

bjarthur commented on src/algorithms/linear_stretching.jl:
````diff
@@ -2,7 +2,7 @@
 ```
     LinearStretching <: AbstractHistogramAdjustmentAlgorithm
-    LinearStretching(; minval = 0, maxval = 1)
+    LinearStretching(; minval = 0, maxval = 1, [src_minval], [src_maxval])
````
for symmetry, maybe `minval` and `maxval` should be `dst_minval` and `dst_maxval` ?

johnnychen94 replied:
I've thought about that once a while, then I think it's okay to keep the previous `minval, maxval` name since those are what most user needs.

And I guess adding another deprecation after `imadjustintensity ` is kind of annoying to users...

bjarthur replied:
i'm not sure about "most users needs".  i for one would only ever need to set `src_{min,max}val`.  worth noting that matlab's [equivalent function](https://www.mathworks.com/help/images/ref/imadjust.html) uses symmetrically named keyword args (`{low,high}_{in,out}`).  i would also argue that naming them `dst_{min,max}val` would be more self-descriptive and hence act as documentation, both in the user-facing API and in the code.

johnnychen94 replied:
Yes, I agree and the re-naming makes sense to me as well. I was just concerning how welcome/unwelcome this renaming will be.

I'm open to changes, but I'd like to first see how @zygmuntszpak thinks before I put my hands on it.

zygmuntszpak replied:
I agree that it is best to rename it to `dst_minval` and `dst_maxval`. Presumably, this means that we will have to amend the version number to become `0.4.0` to reflect the breaking change?

johnnychen94 replied:
> Presumably, this means that we will have to amend the version number to become 0.4.0 to reflect the breaking change?

If we're on `1.x.y` version, a minor version bump is needed. But since we're on `0.3.4` version, _I think_ a patch version is also accepted.

> When you deprecate part of your public API, you should do two things: (1) update your documentation to let users know about the change, (2) issue a new minor release with the deprecation in place
> https://semver.org/#how-should-i-handle-deprecating-functionality

johnnychen94 replied:
In addition to the deprecation of `minval, maxval` in favor of `dest_minval, dest_maxval` in a198016, I also introduced a new set of API in 5674568

* specify (dest_minval, dest_maxval): `LinearStretching(nothing=>(0, 1))`
* specify (src_minval, src_maxval): `LinearStretching((0.1, 0.9))`
* specify both: `LinearStretching((0.1, 0.9) => (0, 1))`

zygmuntszpak reviewed (commented):
Thanks so much for taking this up Johnny, I appreciate your time on it.

johnnychen94 committed:
more tweaks

johnnychen94 committed:
check LinearStretching parameters at construction

johnnychen94 committed:
add convinient constructors for LinearStretching

* LinearStretching((0.1, 0.8)=>(0.2, 0.9))
* LinearStretching(nothing=>(0.2, 0.9))
* LinearStretching((0.1, 0.8))

johnnychen94 committed:
deprecate (minval, maxval) in favor of (dest_minval, dest_maxval)

Although two new fields are introduced to LinearStretching, it doesn't
hurt the general performance; the construction time is about 3ns

johnnychen94 commented:
Hmmm, `Gray{Float32}` performance regression on `Gray{Float32}` is really unexpected 🤔

johnnychen94 committed:
fix Gray{Float32} performance regression

johnnychen94 commented on src/algorithms/linear_stretching.jl:
```diff
@@ -39,44 +43,132 @@ channel are stretched to the specified range. The modified Y channel is then
-            return  T <: Integer ? round(Int, newval ) : newval
+            newval = r * val - o
+            do_clamp && (newval = clamp(newval, dest_minval, dest_maxval))
+            out[p] = T <: Integer ? round(Int, newval) : newval
```
For `Gray{Float32}` input, 47b127b avoids large memory allocation in this line by making `newval` to be `Gray{Float32}` instead of `Float32`

The time difference is __20 ms__ vs __1 ms__

@kimikage do you happen to know what's the issue here? 🤔 why `Gray(0.1) - 0 === 0.1` instead of `Gray(0.1)` (given `Gray(0.1) * 1 === Gray(0.1)`)?

johnnychen94 replied:
I think I've figured it out via `@code_warntype`; that's caused by unstable type inference.

johnnychen94 committed:
the "last" performance tweak

johnnychen94 commented:
Benchmark after 65c7c22:

* the N0f8 version is faster: from __500 μs__ to __260 μs__
* float version isn't changed
* performances are almost the same with/without `clamp`: previously the clamp version is ~__200 μs__ slower for `N0f8`

```julia
julia> img = testimage("mandril_gray");

julia> alg = LinearStretching();
julia> @btime adjust_histogram($img, $alg);
  265.302 μs (6 allocations: 1.25 MiB)
julia> @btime adjust_histogram($(float.(img)), $alg);
  909.417 μs (2 allocations: 1.00 MiB)
julia> @btime adjust_histogram($(Float32.(img)), $alg);
  879.758 μs (2 allocations: 1.00 MiB)

julia> alg = LinearStretching((0.1f0, 0.9f0));
julia> @btime adjust_histogram($img, $alg);
  275.747 μs (6 allocations: 1.25 MiB)
julia> @btime adjust_histogram($(float.(img)), $alg);
  882.314 μs (2 allocations: 1.00 MiB)
julia> @btime adjust_histogram($(Float32.(img)), $alg);
  880.484 μs (2 allocations: 1.00 MiB)
```

I think it's ready to merge (with squash) if there aren't more suggestions.

The color version might need more tweaks ~(e.g., don't convert to YIQ space for RGB images)~ , there might be more optimization space for the float version... but I'll leave it to other PRs.

As a reference, here's the performance of `imadjustintensity`:

```julia
julia> img = testimage("mandril_gray");
julia> @btime imadjustintensity($img);
  1.146 ms (4 allocations: 256.20 KiB)
julia> @btime imadjustintensity($(float.(img)));
  1.175 ms (4 allocations: 1.00 MiB)
julia> @btime imadjustintensity($img, (0.1, 0.9));
  605.017 μs (3 allocations: 256.17 KiB)
julia> @btime imadjustintensity($(float.(img)), (0.1, 0.9));
  341.675 μs (3 allocations: 1.00 MiB) 👀 
```

travis CI failures on macos are unrelated.

zygmuntszpak commented on src/algorithms/linear_stretching.jl:
```diff
@@ -2,13 +2,17 @@
+    LinearStretching(; [src_minval], [src_maxval], dest_minval = 0, dest_maxval = 1)
+
+    LinearStretching((src_minval, src_maxval) => (dest_minval, dest_maxval))
+    LinearStretching((src_minval, src_maxval))
```
Instead of
```julia
 LinearStretching((src_minval, src_maxval))
```

 I would prefer if we made the API 
```julia
LinearStretching((src_minval, src_maxval) => nothing)
```
in accordance with
```julia
LinearStretching(nothing => (dst_minval, dst_maxval)).
```

My concern is that when reading code that uses the `LinearStretching` operation one may mistakenly assume that 

```julia
LinearStretching((0.2, 0.8))
```
specifies the destination interval and not the source interval, especially since the default operaton `LinearStretching()` acts on the destination interval.

johnnychen94 replied:
changed in b3dc554

zygmuntszpak commented on src/algorithms/linear_stretching.jl:
````diff
@@ -2,13 +2,17 @@
 ```
 
-Returns an image where the range of the intensities spans the interval [`minval`, `maxval`].
+Returns an image where the range of the intensities spans the interval [`dest_minval`, `dest_maxval`].
````
I suggest we use `dst_minval` and `dst_maxval` instead of `dest_minval` and `dest_maxval` everywhere. The abbreviation `src` is based on removing the vowels from `source` and taking the first three remaining letters. Hence, it feels more appropriate to keep the three-letter abbreviation `dst` for the sake of symmetry.

johnnychen94 replied:
changed in a703029

zygmuntszpak commented on src/algorithms/linear_stretching.jl:
```diff
@@ -39,44 +43,131 @@ channel are stretched to the specified range. The modified Y channel is then
-using ImageContrastAdjustment, ImageView, TestImages
+using ImageContrastAdjustment, TestImages
 
 img = testimage("mandril_gray")
```
```julia
# Stretches the contrast in `img` so that it spans the unit interval.
```

zygmuntszpak reviewed (commented):
Thanks again for the excellent work, and thank you very much for installing the benchmark tool as well.

johnnychen94 committed:
rename dest_ to dst_

johnnychen94 committed:
use LinearStretching((src_minval, src_maxval)=>nothing) notation

johnnychen94 committed:
clean up performance tweaks and comments

johnnychen94 committed:
correct the constructor signature

cont. b3dc554

johnnychen94 committed:
do clamp after linear stretching instead of before

although clamp before linear stretching is slightly faster, it has
boundary rounding issues, which causes test failures and isn't reliable
to use

bjarthur commented on src/algorithms/linear_stretching.jl:
```diff
@@ -39,44 +43,149 @@ channel are stretched to the specified range. The modified Y channel is then
+    # this is only used when user specifies custom `(src_minval, src_maxval)`
+    out_minval = r * img_min - o
+    out_maxval = r * img_max - o
+    do_clamp = (out_minval < dst_minval) || (out_maxval > dst_maxval)
```
this PR is great as is, and is all i (at least currently) need.  however, making the clamp optional would provide for maximum flexibility.  would this be possible with a keyword argument, or otherwise?

thanks for all the hard work @johnnychen94

johnnychen94 replied:
In the first two commits, it's hard to handle this correctly wrt `FixedPoint`, and then I totally forget about this. Sorry about that.

I've added a `no_clamp` option in a254af4

johnnychen94 committed:
expose no_clamp option to the constructor

johnnychen94 commented:
I didn't expect this PR going so long, so here's a summary:

* expose `src_minval` and `src_maxval` (fixes #27 )
* deprecate `minval`/`maxval` in favor of `dst_minval`/`dst_maxval`
* introduce a new convenient API: `LinearStretching((src_minval, src_maxval) => (dst_minval, dst_maxval))`
* correctly handles the clamp on output intensity
* expose `no_clamp` option to disable the clamp to `[dst_minval, dst_maxval]`: `FixedPoint` inputs are still clamped wrt `[typemin(T), typemax(T)]` to avoid ArgumentError
* performance tweak: 2x-20x performance boost

I plan to merge this in two days if there are no further comments

johnnychen94 commented:
@zygmuntszpak @bjarthur thanks for the insightful suggestions on the API

timholy commented:
Nice work @johnnychen94!

bjarthur commented:
indeed, great work!

sorry i didn't think to suggest these earlier, but two things:

1. it would be convenient to be able to specify a vector of length 2, in addition to a tuple, as the {src,dst}_{min,max}val .  this signature would also be consistent with how the `dims` keyword argument on many julia functions works (e.g. `mean`).

2. how about `clamp=true` instead of `no_clamp=false`?  i know it might seem spurious, but it is not only three fewer letters, but avoids a double negative.  the latter in particular makes me have to think a little harder to not get the sign flipped.

i can submit a PR if there is a consensus about either one or the other or both.

johnnychen94 commented:
Although it doesn't make a real performance difference for `LinearStretching`, in general, accepting vector seems like an anti-pattern to me since tuple is faster than vector. A permanent deprecation or a helpful error message(ref: https://github.com/JuliaLang/julia/pull/35094) might be helpful in this case.

I'm okay to rename it to `clamp`, but perhaps `do_clamp` is more intuitive for a boolean value?

zygmuntszpak commented:
@bjarthur Could you please elaborate on how specifying a vector rather than a tuple is more convenient? I would prefer not to support two different ways of specifying the same thing. 

`do_clamp` sounds odd to me since I read `clamp` as an action (verb) already, and so the `do` seems unnecessary.  Hence, I would probably go with `clamp`. An alternative, which I anticipate people not to like because it is longer, is to have something like `clamp_result = true`.

bjarthur commented:
i believe in julia it is standard to accept both 2-tuples and 2-vectors when specifying the dimensions along which to perform an algebraic operation.  so for example with `mean`:

```
julia> using Statistics

julia> A=rand(3,4,5);

julia> mean(A, dims=(1,3))
1×4×1 Array{Float64,3}:
[:, :, 1] =
 0.394565  0.477098  0.5762  0.563391

julia> mean(A, dims=[1,3])
1×4×1 Array{Float64,3}:
[:, :, 1] =
 0.394565  0.477098  0.5762  0.563391
```

similarly for `quantile` even:

```
julia> V=rand(100);

julia> quantile(V, (0.1,0.9))
(0.07442527144546038, 0.9287217120956874)

julia> quantile(V, [0.1,0.9])
2-element Array{Float64,1}:
 0.07442527144546038
 0.9287217120956874
```

with LinearStretching in some cases the min/max vals you want to use might be calculated and stored in a vector and it is a hassle for the user to `tuple` them.  particularly when supporting vectors i think is as simple as adding the following method:

```
LinearStretching(rangemap::Pair{Vector, Vector}; no_clamp=false) =
    LinearStretching(rangemap.first..., rangemap.second..., nothing, nothing, no_clamp)
```

johnnychen94 commented:
> with LinearStretching in some cases the min/max vals you want to use might be calculated and stored in a vector and it is a hassle for the user to tuple them.

That could be quite a fair use case for me. I think it's okay to get this since I believe it doesn't get things worse in this case. However, I'm not excited to add this to the codebase; I kind of prefer to distinguish tuple and vector and to not mixing them.

zygmuntszpak commented:
@bjarthur Perhaps you could open an issue on the Images repo regarding your suggestion of also allowing one to pass a vector instead of a tuple. If we do this here, then I would like it to become a consensus for the Images ecosystem. I want to avoid a situation where some key functions do this kind of conversion automatically, and others not.
